bpfd/src/storage.rs

```diff
@@ -1,209 +1,371 @@
+        info!(
+            "Received unpublish volume request with :\n\
+        volume_id: {volume_id},\n\
+        target_path: {target_path}"
+        );
+
+        Ok(Response::new(NodeUnpublishVolumeResponse {}))
     }
     async fn node_get_volume_stats(
         &self,
         _request: Request<NodeGetVolumeStatsRequest>,
     ) -> std::result::Result<Response<NodeGetVolumeStatsResponse>, tonic::Status> {
-        todo!()
+        return Err(Status::unimplemented(""));
     }
     async fn node_expand_volume(
         &self,
         _request: Request<NodeExpandVolumeRequest>,
     ) -> std::result::Result<Response<NodeExpandVolumeResponse>, tonic::Status> {
-        todo!()
+        return Err(Status::unimplemented(""));
     }
     async fn node_get_capabilities(
         &self,
         _request: Request<NodeGetCapabilitiesRequest>,
     ) -> std::result::Result<Response<NodeGetCapabilitiesResponse>, tonic::Status> {
-        todo!()
+        return Ok(Response::new(NodeGetCapabilitiesResponse {
+            capabilities: vec![NodeServiceCapability {
+                r#type: Some(node_service_capability::Type::Rpc(
+                    node_service_capability::Rpc {
+                        r#type: node_service_capability::rpc::Type::Unknown.into(),
+                    },
+                )),
+            }],
+        }));
     }
     async fn node_get_info(
         &self,
         _request: Request<NodeGetInfoRequest>,
     ) -> std::result::Result<Response<NodeGetInfoResponse>, tonic::Status> {
-        todo!()
+        return Ok(Response::new(NodeGetInfoResponse {
+            node_id: self.driver.clone().node_id,
+            max_volumes_per_node: 0,
+            accessible_topology: None,
+        }));
     }
 }
 
 impl StorageManager {
-    pub fn new() -> Self {
-        let csi_controller = CsiController {};
-        let csi_node = CsiNode {};
+    // TODO add endpoint configuration
+    pub fn new(name: String, node_id: String, version: String) -> Self {
+        let driver = StorageDriver {
+            name,
+            node_id,
+            version,
+            cap: vec![ControllerServiceCapability {
+                r#type: Some(controller_service_capability::Type::Rpc(
+                    controller_service_capability::Rpc {
+                        r#type: controller_service_capability::rpc::Type::Unknown.into(),
+                    },
+                )),
+            }],
+            _vc: vec![volume_capability::AccessMode {
+                mode: volume_capability::access_mode::Mode::SingleNodeWriter.into(),
+            }],
+        };
+        let csi_identity = CsiIdentity {
+            driver: driver.clone(),
+        };
+        let csi_controller = CsiController {
+            driver: driver.clone(),
+        };
+        let csi_node = CsiNode { driver };
         Self {
             csi_controller,
             csi_node,
+            csi_identity,
         }
     }
 
     pub async fn run(self) {
```

Comment:
Is this mandatory in the spec? What if something is already running on that socket?

Reply:
Actually this is just mounted in the bpfd pod at /run/bpfd/sock/csi.sock.

on the host it will be mounted in the predefined `/var/lib/kubelet/plugins/csi-bpfd` so we will always own the socket there i.e we don't care if something else is running there.

Reply:
Additionally I moved the default bpfd socket to the runtime dir as well rather then the state dir `/var/lib`